### PR TITLE
feat(react): add indeterminate support to checkbox

### PR DESCRIPTION
**Pull request by chornonoh-vova**
closes: #479

### CODE REVIEW

File: packages/react/src/components/Checkbox/index.tsx [hunk 1]
```diff
@@ -36,23 +36,35 @@ const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
...
     useEffect(() => {
       setIsChecked(checked);
     }, [checked]);
 
     const refProp = ref || checkboxRef;
-    if (typeof refProp === 'function') {
-      refProp(checkRef.current);
-    }
+
+    const refCallback = (elem: HTMLInputElement | null): void => {
```
Comment (scurker) [inline packages/react/src/components/Checkbox/index.tsx:55]: Is there a reason this ref callback function was added?

If you're trying to access the ref, `useEffect` may be what you want here.

File: packages/react/src/components/Checkbox/index.tsx [hunk 2]
```diff
@@ -92,6 +110,9 @@ const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
                 onBlur(e);
               }
             }}
+            aria-checked={
```
Comment (scurker) [inline packages/react/src/components/Checkbox/index.tsx:113]: I'm not sure this is necessary. 

Looking at the accessibility tree, a checkbox with the value of `indeterminate` already has the right properties:

chrome accessibility tree showing a value of checked: mixed

File: packages/styles/forms.css [hunk 3]
```diff
@@ -20,6 +20,11 @@
   --field-icon-inactive-color: var(--gray-90);
   --field-icon-active-color: rgba(60, 122, 174, 0.25);
   --field-icon-error-active-color: rgba(217, 50, 81, 0.25);
+  --field-icon-indeterminate-color: var(--accent-primary);
+  --field-icon-indeterminate-disabled-color: var(
+    --accent-primary-dsiabled,
+    #78a6d8
+  );
```
Comment (scurker) [inline packages/styles/forms.css:27]: I don't think new colors are necessary. This should be using the existing `--field-icon-xxx` colors for the appropriate states.

File: packages/react/src/components/Icon/icons/checkbox-indeterminate.svg
Comment (scurker) [inline packages/react/src/components/Icon/icons/checkbox-indeterminate.svg:1]: A new icon is not necessary, please take a look at an example of how this should be styled in the component proposal.

File: packages/react/src/components/Checkbox/Checkbox.test.tsx [hunk 4]
```diff
@@ -46,6 +46,29 @@ test('should render disabled checked checkbox', () => {
...
 });
 
+test('should render indeterminate checkbox', () => {
+  const ref = createRef<HTMLInputElement>();
+  render(<Checkbox id="id" label="checkbox" ref={ref} indeterminate />);
+  expect(ref.current).toBeInstanceOf(HTMLInputElement);
+  expect(ref.current?.indeterminate).toBeTruthy();
+  expect(ref.current).toEqual(
+    screen.queryByRole('checkbox', { name: 'checkbox' })
+  );
```
Comment (scurker) [inline packages/react/src/components/Checkbox/Checkbox.test.tsx:56]: This isn't quite how we should be doing the assertions here. We should be limiting to just what should be tested for this specific test, i.e. that an indeterminate checkbox was rendered. The way this test is written we're testing that the ref value has an indeterminate value but not that a indeterminate value was rendered on the page. Please take a look at how some of the other tests are written here to try to eliminate down to only the assertions that are necessary.

Additionally, jest-dom already has a matcher we can use, [`toBePartiallyChecked()`](https://github.com/testing-library/jest-dom?tab=readme-ov-file#tobepartiallychecked)

File: packages/react/src/components/Checkbox/Checkbox.test.tsx
Comment (scurker) [inline packages/react/src/components/Checkbox/Checkbox.test.tsx:1]: We need to include the `indeterminate` state as part of the accessibility checks as well.

**Review by scurker — CHANGES_REQUESTED** [review]
This is a great start! 

As an overview for testing, please take a look at the [testing strategies](https://github.com/dequelabs/cauldron/blob/develop/CONTRIBUTING.md#testing-strategies) guidelines as there's a few things we're missing.

Additionally, it's not documented (yet) but we're capturing screenshots for all interactive elements. We should include the `indeterminate` checkbox as an additional state to capture here: https://github.com/dequelabs/cauldron/blob/develop/packages/react/src/components/Checkbox/screenshots.e2e.tsx

**Comment by scurker** [comment]
One additional thing that I'm realizing is we're not matching what native checkboxes do. 

A checkbox can have an `indeterminate` state and be `checked` / `unchecked`. When clicked, the indeterminate state is removed and the `checked` value is flipped from it's currently value so we will likely want to match that behavior as our current checkbox component isn't "controlled".

File: packages/react/src/components/Checkbox/index.tsx (hunk 1 above)
Comment (chornonoh-vova) [reply packages/react/src/components/Checkbox/index.tsx]: Thanks, removed

File: packages/react/src/components/Checkbox/index.tsx (hunk 2 above)
Comment (chornonoh-vova) [reply packages/react/src/components/Checkbox/index.tsx]: Thanks, removed

File: packages/styles/forms.css (hunk 3 above)
Comment (chornonoh-vova) [reply packages/styles/forms.css]: Thanks, removed

File: packages/react/src/components/Icon/icons/checkbox-indeterminate.svg
Comment (chornonoh-vova) [reply packages/react/src/components/Icon/icons/checkbox-indeterminate.svg:1]: Thanks, implemented without icon

File: packages/react/src/components/Checkbox/Checkbox.test.tsx (hunk 4 above)
Comment (chornonoh-vova) [reply packages/react/src/components/Checkbox/Checkbox.test.tsx]: Thanks, updated tests

File: packages/react/src/components/Checkbox/Checkbox.test.tsx
Comment (chornonoh-vova) [reply packages/react/src/components/Checkbox/Checkbox.test.tsx:1]: Thanks, updated a11y tests

**Review by scurker — APPROVED** [review]
At a high level, this looks good - thanks for making changes. I will give it a final pass over tomorrow morning.

File: packages/react/src/components/Checkbox/index.tsx [hunk 5]
```diff
@@ -1,15 +1,14 @@
...
-  Ref,
   useState,
   useEffect,
   useRef,
   useMemo
 } from 'react';
 import classNames from 'classnames';
 import nextId from 'react-id-generator';
-import Icon from '../Icon';
+import Icon, { IconType } from '../Icon';
```
Comment (scurker) [inline packages/react/src/components/Checkbox/index.tsx:11]:
```suggestion
import Icon, { type IconType } from '../Icon';
```

**Review by scurker — COMMENTED** [review]
Very very minor nit, otherwise looks good!

**Comment by shawnsharpDQ** [comment]
Is there anything we can do for the visibility of the disabled indeterminate state? 
image
compared to the disabled checked 
image

**Review by shawnsharpDQ — DISMISSED** [review]

**Review by scurker — APPROVED** [review]

**Pull request merged**